Swiftz/Reader.swift

```diff
@@ -0,0 +1,174 @@
+//
+//  Reader.swift
+//  Swiftz
+//
+//  Created by Matthew Purland on 11/25/15.
+//  Copyright © 2015 TypeLift. All rights reserved.
+//
+
+/// A `Reader` monad with `R` for environment and `A` to represent the modified environment.
+public struct Reader<R, A> {
+    /// The function that modifies the environment
+    public let reader : R -> A
+    
+    init(_ reader : R -> A) {
+        self.reader = reader
+    }
+    
+    /// Transform the value returned by a reader
+    public func mapReader<B>(f : A -> B) -> Reader<R, B> {
+        return Reader<R, B>(f • runReader)
+    }
+    
+    /// Runs the reader and extracts the final value from it
+    public func runReader(environment : R) -> A {
+        return reader(environment)
+    }
+    
+    /// Executes a computation in a modified environment
+    public func local(f : R -> R) -> Reader<R, A> {
+        return Reader(reader • f)
+    }
+
+    /// Retrieves the monad environment
+    public func ask() -> R -> A {
+        return reader
+    }
+}
+
+/// Puts a value in a reader.
+public func returnReader<R, A>(a : A) -> Reader<R, A> {
+    return Reader<R, A>.pure(a)
+}
+
+/// Runs the reader and extracts the final value from it. This provides a global function for running a reader.
+public func runReader<R, A>(reader : R -> A) -> R -> A {
+    return runReader(Reader<R, A>(reader))
+}
+
+public func runReader<R, A>(reader : () -> R -> A) -> R -> A {
+    return runReader(reader())
+}
+
+public func runReader<R, A>(reader : Reader<R, A>) -> R -> A {
+    return reader.runReader
+}
+
+/// Retrieves the monad environment
+public func ask<A>() -> A -> A {
```

Comment:
I've never seen this particular overloading before.  What's the expected use case?